tutorials/llm/llama/domain-adaptive-pretraining/code/custom_tokenization.ipynb

```diff
@@ -0,0 +1,2569 @@
+     ]
+    }
+   ],
+   "source": [
+    "data_root = \"./data/all_jsonl_data_sample/\"       # path where the domain specific data is stored\n",
+    "save_root = \"./models/tokenizer/llama2/\"    # path to save the finetuned opt tokenizer\n",
+    "batch_size = 1000    # batch size used in the tokenization process\n",
+    "vocab_size = 20000   # target vocab size for training opt tokenizer\n",
+    "\n",
+    "# ensure that the directory exists before changing permissions\n",
+    "directory = \"../code/\"\n",
+    "is_directory = os.path.isdir(directory)\n",
+    "print(f\"Is a directory: {is_directory}\")\n",
+    "\n",
+    "# change permissions to ensure we have read, write and execute permissions\n",
+    "! chmod ugo+rwx ../code/"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "519e6c13-45c3-4f52-9ed0-770d4ec62766",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1150: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.\n",
+      "  warnings.warn(\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Before Training: \n",
+      "total token cnt 66025\n",
+      "\n",
+      "\n",
+      "\n",
+      "After Training: \n",
+      "total token cnt 47712\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "('./models/tokenizer/llama2/custom_tokenizer_init_20000.json/tokenizer_config.json',\n",
+       " './models/tokenizer/llama2/custom_tokenizer_init_20000.json/special_tokens_map.json',\n",
+       " './models/tokenizer/llama2/custom_tokenizer_init_20000.json/vocab.json',\n",
+       " './models/tokenizer/llama2/custom_tokenizer_init_20000.json/merges.txt',\n",
+       " './models/tokenizer/llama2/custom_tokenizer_init_20000.json/added_tokens.json',\n",
+       " './models/tokenizer/llama2/custom_tokenizer_init_20000.json/tokenizer.json')"
+      ]
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Train a tokenizer from scratch and save output files\n",
+    "keys = [\"text\"] # keys to extract from json files\n",
+    "tokenizer = AutoTokenizer.from_pretrained(\"facebook/opt-350m\") # load pre-trained tokenizer (https://huggingface.co/facebook/opt-350m)\n",
+    "# Train the tokenizer from scratch on a new corpus with the same defaults (in terms of special tokens or tokenization pipeline) as the current one.\n",
+    "tokenizer = train_tokenizer(data_root, batch_size, vocab_size, tokenizer, keys)\n",
+    "\n",
+    "#Save and print paths\n",
+    "tokenizer.save_pretrained(save_root + \"custom_tokenizer_init_\" + str(vocab_size) + \".json\")"
```

Comment:
The save directory is named 'custom_tokenizer_init_x.json', suggest changing the directory name to not end with .json.